### PR TITLE
Fix Slice shape inference in the case of dynamic inputs

### PR DESCRIPTION
**Pull request by AtanasDimitrovQC**
### Description
Slice shape-inference didn't work when the start or stop inputs were not defined. 

### Motivation and Context
This contributes to making the type and shape inference more complete.

### CODE REVIEW

**Comment by justinchuby** [comment]
@onnx/sig-operators-approvers for approval

**Comment by justinchuby** [comment]
Could you fix dco by signing the new commit?

**Comment by adityagoel4512** [comment]
@justinchuby could we please add this to the merge queue?

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -958,6 +958,11 @@ ONNX_OPERATOR_SET_SCHEMA(
 
           if (!startsInitializer || !endsInitializer || (hasInputShape(ctx, 3) && !ctx.getInputData(3)) ||
               (hasInputShape(ctx, 4) && !ctx.getInputData(4))) {
+            const auto input_rank = ctx.getInputType(0)->tensor_type().shape().dim_size();
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:961]: This looks good, thanks! Unrelated to this change, I wonder if we should change the `hasInputShape` in the previous two lines to `hasInput`? We need to only check if the starts/ends inputs are supplied, not whether we know their shape.

**Review by gramalingam — APPROVED** [review]
LGTM. I think we might be able to fix another minor issue at the same time, as mentioned in the comments.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (justinchuby) [reply onnx/defs/tensor/defs.cc:961]: Failed. Reverting

**Pull request merged**